dll/win32/shell32/dialogs/item_prop.cpp

```diff
@@ -0,0 +1,182 @@
+        ILFree(pidl);
+    }
+    return hr;
+}
+
+struct ShellPropSheetDialog
+{
+    typedef void (CALLBACK*PFNINITIALIZE)(LPCWSTR InitString, IDataObject *pDO,
+                                          HKEY *hKeys, UINT *cKeys);
+
+    static HRESULT Show(const CLSID *pClsidDefault, IDataObject *pDO,
+                        PFNINITIALIZE InitFunc, LPCWSTR InitString)
+    {
+        HRESULT hr;
+        CRegKeyHandleArray keys;
+        if (InitFunc)
+            InitFunc(InitString, pDO, keys, keys);
+        WCHAR szCaption[MAX_PATH], *pszCaption = NULL;
+        if (SUCCEEDED(SHELL_GetCaptionFromDataObject(pDO, szCaption, _countof(szCaption))))
+            pszCaption = szCaption;
+        hr = SHOpenPropSheetW(pszCaption, keys, keys, pClsidDefault, pDO, NULL, NULL) ? S_OK : E_FAIL;
+        return hr;
+    }
+
+    struct DATA
+    {
+        PFNINITIALIZE InitFunc;
+        LPWSTR InitString;
+        CLSID ClsidDefault;
+        const CLSID *pClsidDefault;
+        IStream *pObjStream;
+    };
+
+    static void FreeData(DATA *pData)
+    {
+        if (pData->InitString)
+            SHFree(pData->InitString);
+        SHFree(pData);
+    }
+
+    static HRESULT ShowAsync(const CLSID *pClsidDefault, IDataObject *pDO,
+                             PFNINITIALIZE InitFunc, LPCWSTR InitString)
+    {
+        DATA *pData = (DATA*)SHAlloc(sizeof(*pData));
+        if (!pData)
+            return E_OUTOFMEMORY;
+        ZeroMemory(pData, sizeof(*pData));
+        pData->InitFunc = InitFunc;
+        if (InitString && FAILED(SHStrDupW(InitString, &pData->InitString)))
+        {
+            FreeData(pData);
+            return E_OUTOFMEMORY;
+        }
+        if (pClsidDefault)
+        {
+            pData->ClsidDefault = *pClsidDefault;
+            pData->pClsidDefault = &pData->ClsidDefault;
+        }
+
+        HRESULT hr = S_OK;
+        if (pDO)
+            hr = CoMarshalInterThreadInterfaceInStream(IID_IDataObject, pDO, &pData->pObjStream);
+
+        UINT flags = CTF_COINIT | CTF_PROCESS_REF | CTF_INSIST;
+        if (SUCCEEDED(hr) && !SHCreateThread(ShowPropertiesThread, pData, flags, NULL))
+        {
+            if (pData->pObjStream)
+                pData->pObjStream->Release();
+            hr = E_FAIL;
+        }
+        if (FAILED(hr))
+            FreeData(pData);
+        return hr;
+    }
+
+    static DWORD CALLBACK ShowPropertiesThread(LPVOID Param)
+    {
+        DATA *pData = (DATA*) Param;
+        CComPtr<IDataObject> pDO;
+        if (pData->pObjStream)
+            CoGetInterfaceAndReleaseStream(pData->pObjStream, IID_PPV_ARG(IDataObject, &pDO));
+        Show(pData->pClsidDefault, pDO, pData->InitFunc, pData->InitString);
+        FreeData(pData);
+        return 0;
+    }
+};
+
+static void CALLBACK
+FSFolderItemPropDialogInitCallback(LPCWSTR InitString, IDataObject *pDO, HKEY *hKeys, UINT *cKeys)
+{
+    // Add file-type specific pages
```

Comment:
is this a TODO?

Reply:
No, it is a comment about what this callback does but I'll move the comment to the important part of the function.